istioctl/cmd/revision.go

```diff
@@ -0,0 +1,659 @@
+			}
+			if len(args) > 1 {
+				return fmt.Errorf("exactly 1 revision should be specified")
+			}
+			if revArgs.verbose && len(revArgs.sections) != 0 {
+				return fmt.Errorf("sections and verbose flags should not be specified at the same time")
+			}
+			isValidFormat := false
+			for _, f := range validFormats {
+				if revArgs.output == f {
+					isValidFormat = true
+					break
+				}
+			}
+			if !isValidFormat {
+				return fmt.Errorf("unknown format %s. It should be %#v", revArgs.output, validFormats)
+			}
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			revArgs.name = args[0]
+			if revArgs.name != "<default>" {
+				if errs := validation.IsDNS1123Label(revArgs.name); len(errs) > 0 {
+					return fmt.Errorf(strings.Join(errs, "\n"))
+				}
+			} else {
+				revArgs.name = "" // TODO(su225): This case is broken. Fix it.
+			}
+			logger := clog.NewConsoleLogger(cmd.OutOrStdout(), cmd.ErrOrStderr(), scope)
+			return revisionDescription(cmd.OutOrStdout(), &revArgs, logger)
+		},
+	}
+	describeCmd.Flags().BoolVarP(&revArgs.verbose, "verbose", "v", false, "Dump all information related to the revision")
+	describeCmd.Flags().StringSliceVarP(&revArgs.sections, "sections", "s", []string{}, "Sections that should be included in description output")
+	describeCmd.Flags().StringVarP(&revArgs.output, "output", "o", "table", "Output format for revision description")
+	return describeCmd
+}
+
+func revisionListCommand() *cobra.Command {
+	listCmd := &cobra.Command{
+		Use:     "list",
+		Short:   "Show list of control plane and gateway revisions that are currently installed in cluster",
+		Example: `   istioctl experimental revision list`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			logger := clog.NewConsoleLogger(cmd.OutOrStdout(), cmd.ErrOrStderr(), scope)
+			return revisionList(cmd.OutOrStdout(), &revArgs, logger)
+		},
+	}
+	return listCmd
+}
+
+// TODO: Come up with a better name
+type revisionInfo struct {
+	Revision string
+	Tags     []string
+	IOPs     []*iopv1alpha1.IstioOperator
+}
+
+func newRevisionInfo(name string) *revisionInfo {
+	return &revisionInfo{
+		Revision: name,
+		Tags:     []string{},
+		IOPs:     []*iopv1alpha1.IstioOperator{},
+	}
+}
+
+func revisionList(writer io.Writer, args *revisionArgs, logger clog.Logger) error {
+	client, err := newKubeClient(kubeconfig, configContext)
+	if err != nil {
+		return fmt.Errorf("cannot create kubeclient for kubeconfig=%s, context=%s: %v",
+			kubeconfig, configContext, err)
+	}
+
+	revisions := map[string]*revisionInfo{}
+
+	// Get a list of control planes which are installed in remote clusters
+	// In this case, it is possible that they only have webhooks installed.
+	webhooks, err := getWebhooks(context.Background(), client)
+	if err != nil {
+		return fmt.Errorf("error while listing webhooks: %v", err)
+	}
+	for _, hook := range webhooks {
+		rev := renderWithDefault(hook.GetLabels()[label.IstioRev], "default")
```

Comment:
What will happen if the only installed Istio is with revision eg: `istioctl install --revision 1.8.1` and there is no default revision available in the cluster?

Reply:
I came across this situation in one of the PR https://github.com/istio/istio/pull/29442#issuecomment-740372767

Reply:
Did you mean the other way around? That is, only `<default>` revision is installed in the cluster? Yeah. There are issues with default revision. First, I'm trying to get it working for cases where revision label exists (that is, revision is given). Then I'll fix the `<default>` special case. @esnible also pointed out another issue where `describe <default>` does not work. That is also in queue.

Reply:
Just got rid of `<default>` thing altogether. Now simply doing `istioctl x revision describe default` works fine.